### ISSUE TITLE
Support multiple languages

### ISSUE THREAD

**Issue by PoLaKoSz** [tonykaralis/MoodMovies#5]

**Comment by tonykaralis** [comment]
Hello there, Do you have any suggestions?

**Comment by PoLaKoSz** [comment]
Hi!

I am not an experienced developer but I use [ResX Manager](https://marketplace.visualstudio.com/items?itemName=TomEnglert.ResXManager) and [this technique](https://docs.microsoft.com/en-us/dotnet/framework/wpf/advanced/wpf-globalization-and-localization-overview) (x:Uid="<keyName>") to make my app support different languages.

Did I misunderstand Your question?

**Comment by tonykaralis** [comment]
No not all. Thanks for the tip, I will definitely look into ResX Manager. 
Do you have any other ideas, you might like to see in this app?

**Comment by PoLaKoSz** [comment]
This is an example but I personally use [YTS.AM](https://yts.am/browse-movies) to download movies so I think a feature to not just search but download it (if it is a .torrent just open it with the default app) directly would be great. YTS also has a [nice API](https://yts.am/api).

If You want I happily help to develop this project but as I mentioned before I am a junior developer.

**Comment by tonykaralis** [comment]
That sounds like a really good idea. Dont worry about it, I would like the help 👍  Have you done any work with C#, xaml, Caliburn.Micro, material design toolkit, mahApps toolkit and MVVM?

**Comment by PoLaKoSz** [comment]
Yes, I have the fundamentals of C#, XAML and MVVM too but maybe a little feature or something would show my skillset to You.

I was never used Caliburn.Micro or Material design or MahAps but it's not a problem for me to learn new things like this if I have motivation to do it. :smile:

**Comment by tonykaralis** [comment]
Perfect. Caliburn makes MVVM so much easier. Material Design and Mahapps help make everything look good. They are all easy things to learn. Your English is really good, I can understand everything 💯 %.
So where do you want to start? What feature would you like to work on?  :smiley:

**Comment by PoLaKoSz** [comment]
Good to hear that 😄 

I don't feel confident implementing a wrapper around an whole API so I think this multilanguage support is what I need to choose.

But I need your advice how to localize a message like this:
https://github.com/tonykaralis/MoodMovies/blob/d83fcc133f627b66428e2f93a6fb44d36ec2ee28/MoodMovies/ViewModels/SearchViewModel.cs#L63
I don't want to set string in the code behind like this:
```` csharp
eventAgg.PublishOnUIThread(new StartLoadingMessage(new ResourceLoader().GetString("SearchView_SearchInProgress")));
````
Is there a way to pass a XAML document or something like this instead where I can use the x:Uid directive?

**Comment by tonykaralis** [comment]
No problem, multi language support sounds great.  Check this post out [Marked Question](https://stackoverflow.com/questions/11327840/multilingual-wpf-application), this could work well for creating seperate resource files and loading the one we want based on users choice during runtime.  This solution will work for static strings in the xaml code. However, I am not sure yet how to work with strings that need to be displayed from the ViewModel(dynamically changing). As soon as i get some time i will look into this, but if you want to get started then please do 😄

**Comment by PoLaKoSz** [comment]
Thank You for the suggestion but I wasn't be able to make it work.

Currently I try to [build this library](https://www.codeproject.com/Articles/35159/WPF-Localization-Using-RESX-Files) to the project.

**Comment by tonykaralis** [comment]
Cool, let me know how it goes 👍

**Comment by PoLaKoSz** [comment]
You can find the working **demo** in my [fork](https://github.com/PoLaKoSz/MoodMovies/tree/feature/globalization-Infralution.Localization.Wpf). Yes, I know this is ugly I made this just to demonstrate how it works. Let me know your thoughts.

It's in the [description of the library](https://www.codeproject.com/Articles/35159/WPF-Localization-Using-RESX-Files) but I think I mention it:
> (...) store the resources associated with each WPF window or control in a separate RESX file (typically named the same as the Window). This is important for projects with a large number of windows where it can become difficult to manage the large number of resources in a single file, particularly with multiple developers. It also means that your resource names only have to be unique within the window.

![File integration](https://user-images.githubusercontent.com/1327533/40417768-ddb471a6-5e80-11e8-9d22-2eaa427d66db.PNG)

**Comment by tonykaralis** [comment]
I like it 👍 Could you upload this file C:\Users\PoLaKoSz\Downloads\wpflocalization\Infralution.Localization.Wpf\Infralution.Localization.Wpf.Lib\Infralution.Localization.Wpf.csproj  or include it to your fork, so I can run it and see how it works. I think your idea will work!

**Comment by PoLaKoSz** [comment]
Damn! I roasted myself. Jesus

I added a commit to fix this issue and made a stap forward to localize the UI. I'm disappointed because I found 2 bugs and I don't have a clue where to fix it.
- When you run the app and open the navigation bar the menus not localized but when you change the languge in the ListView the menu localize itself
- The search button style is missing or something wrong with it (image not shown)

**Comment by tonykaralis** [comment]
Very nice. I don't understand the first bug, the second bug I will look at this ASAP.

**Comment by PoLaKoSz** [comment]
Sorry. My habit is to overcomplicate my sentences. So 
- Run the app
- Open the hamburger `menu`
- I see this: 
![capture](https://user-images.githubusercontent.com/1327533/40424881-12facc82-5e97-11e8-9dcd-a43619bfb90a.PNG)
- Close the Hamburger menu
- Change the language to `magyar (Magyarország)`
- And now I see this:
![capture](https://user-images.githubusercontent.com/1327533/40424947-464c9c32-5e97-11e8-884c-8d6f8a256b73.PNG)
- If I change the language again to `English` I see the the menu names as it should be in the first place:
![capture](https://user-images.githubusercontent.com/1327533/40425016-741074e0-5e97-11e8-89d0-ecc8b823ca22.PNG)

**Comment by tonykaralis** [comment]
Oh ok, I see.
### Solution to bug 1:
The problem is in that MainView I am using markup within markup(MaterialDesign Toolkit) which means we will need to set the resname manually for each item on the mainviewmodel.  
See note:  

> Setting the DefaultResxName using the attached property works provided you are using the Resx element within a normal FrameworkElement property. It will not work however when the Resx element is used within another MarkupExtension. In this case, you will still need to set the ResxName explicitly.
> 
To fix this, all we need to do is include the full resx name for each control containing a resx string:
eg.
`    Line 39 Content="{Resx ResxName=MoodMovies.Views.MainView ,Key =_menuName_Search}"` 

#### Solution to another bug I found
Also I noticed that the designer is throwing an exception which is not allowing us to view the designer in design time. To fix replace this `    Language="{UICulture}"` with this `    Language="{Binding Path=UICulture}"` in all the views that are using resx.
I hope this helps. 
I am not sure what is wrong with the button image yet.

**Comment by tonykaralis** [comment]
Yeah i think maybe the resx library is interfering with the icon. That projects looks like it could b interesting.
keep me updated 👍 
Nice fix 👍

**Comment by PoLaKoSz** [comment]
I hope I have a great news: I [created a new branch](https://github.com/PoLaKoSz/MoodMovies/tree/feature/globalization-WPFLocalizationExtension) with the mentioned package and it looks like to me it working without any bug.

*I localized only the Search menu name*

**Comment by tonykaralis** [comment]
This is good news. No bugs, translation happens quick and the button icon is ok. I like this package. 👍 
Well done, this is impressive.

**Comment by PoLaKoSz** [comment]
Good to hear.

I think You should make a convention how should I name the string keys in the `.rsx` files to follow some convention what do You think about it?

**Comment by tonykaralis** [comment]
Yes I agree. 
#### 1. I will be making a few changes tonight to the UI. These include string modfications. 
#### 2. Once I have done those changes, then _you can pull the changes I made into your fork_. 
#### 3. After this you can finish the localization.
#### 4. When you are finished localization, make a pull request.
#### 5. I will accept the request and merge. And then I will move the language selector.(I still need to think where i want to put it)

If you are happy with this then we have a plan.

I have made a lot of changes today 😃 

#### With regards to naming conventions for resx string:
What do you think about the following?
`    <ViewName>_<control>_<textToTranslate>`
eg. `    SearchView_TextBox_Actor` or `    Main_Button_Search` 
This way we now where the control is, what it is, and what the text is to translate?
Let me know what your thoughts are. 👍

**Comment by PoLaKoSz** [comment]
Sounds really good I love it. The convention is great. :smile:

**Comment by tonykaralis** [comment]
I have pushed a whole set of changes to the master branch. feel free to pull form upstream whenever you wish. there is a lot of database stuff included which might require you to install some sqlite stuff to get it to work. if you have a problem let me know.
[Guide to updating your fork](https://stackoverflow.com/questions/7244321/how-do-i-update-a-github-forked-repository)
I also added the combobox with the languages to the top right corner and set up some values. I left some notes around for you to replace whatever you need with your language localization code. 👍

**Comment by PoLaKoSz** [comment]
Thank You!

I don't really understand what do You mean by the `I feel ...` ComboBox. This is some kind of advanced search by genre? Maybe it's a dumb question but I don't know how to translate it to my mother language.

| UI       | What I guess it means |
|----------|-----------------------|
| Happy    |               Comedy |
| Thrilled |      Action or Drama |
| Scared   |               Horror |
| Loved    |             Romantic |

**Comment by tonykaralis** [comment]
Ok so the 'I feel.. combobox is where the user will choose would mood they want to be in. So if they want to feel happy or sad or scared.. I am designing a complex algorithm that will map various genres to the different 'moods' that are available. So instead of giving them option to choose genre, instead they choose the mood they want 😄  Something different to all other applications, thats why the name is Moods and Movies.  

> But It's kind of interesting because I feel .... scared and (s)he gets Horror movies to sh@t in his / her pants? 

This is exactly correct. Haha.

So the best way is to translate the exact words I have used, not the genres 👍

**Comment by PoLaKoSz** [comment]
Good Morning!

Can You check [this branch ](https://github.com/PoLaKoSz/MoodMovies/tree/feature/globalization) for any missing localization?

**Comment by tonykaralis** [comment]
Good morning, It looks like all the localization is done so far. Its looking good.
I think for now localization is done and we can close this issue and you can make a pull request.If you agree, let me know and Ill close this.
